### DISCUSSION TITLE
"Handling Pylance Error When Using Different Argument Names in Custom __init__ with Pydantic"

### DISCUSSION THREAD

**Discussion by takerfume** [pydantic/pydantic#8371] [General]
Hello,

I'm encountering an issue with Pylance when I use custom `__init__` methods in Pydantic models where the argument names are different from the field names. The code runs without errors, but Pylance reports the following issue:

```
No parameter named "cpu_prepro_parent_dir"PylancereportGeneralTypeIssues
(function) cpu_prepro_parent_dir: Unknown
```

Here is my code snippet:

```python
from .base import BaseConfig

class LocalDataParentDirConfig(BaseConfig):
    cpu_prepro_parent_dir: str
    gpu_prepro_parent_dir: str
    vslam_parent_dir: str
    postpro_parent_dir: str

    def __init__(self, local_data_parent_dir: str):
        cpu_prepro_parent_dir: str = f"{local_data_parent_dir}/cpu_prepro"
        gpu_prepro_parent_dir: str = f"{local_data_parent_dir}/gpu_prepro"
        vslam_parent_dir: str = f"{local_data_parent_dir}/vslam"
        postpro_parent_dir: str = f"{local_data_parent_dir}/postpro"
        super().__init__(
            cpu_prepro_parent_dir=cpu_prepro_parent_dir,
            gpu_prepro_parent_dir=gpu_prepro_parent_dir,
            vslam_parent_dir=vslam_parent_dir,
            postpro_parent_dir=postpro_parent_dir,
        )
```

How can I fix this type error.

**Comment by brianm78** [comment]
I think the code you give *should* work, if you're constructing a LocalDataParentDirConfig directly.

However, I ran into this post while googling an issue that I suspect may be what you're actually hitting, in that you'll get errors like that for any **subclasses** of your base class, so commenting here for anyone else following google here.

The reason this happens is because they way type checkers detect the "dataclass-like" behaviour of pydantic etc is through the dataclass_transform decorator as defined by  [PEP681](https://peps.python.org/pep-0681/) , which notably contains the following (emphasis added):

>> As with dataclass, method synthesis is skipped if it would overwrite a method that is explicitly declared within the class. **Method declarations on base classes do not cause method synthesis to be skipped.**
>> For example, if a class declares an __init__ method explicitly, an __init__ method will not be synthesized for that class.

Ie. an `__init__` on your base class will not be treated as being inherited by a subclass, but rather it'll go back to the "dataclass-like" behaviour of expecting an auto-generated constructor.  I assume this rule is there so stuff like dataclasses / pydantic's custom __init__ that does the work don't immediately override this, but it does make things a bit awkward if you want to augment / change how `__init__` works for subclasses.

The only workaround I've found is to explicitly do:

    class SubClass(YourBaseClassWithCustomInit):
        __init__ = YourBaseClassWithCustomInit.__init__

In every subclass, which is less than ideal, so it may be a better idea to avoid this altogether, and do any such customisation as extra factory classmethods.
